tests/helpers.ts

```diff
@@ -39,3 +41,48 @@ export async function advanceTime({
 export async function flushPromises(): Promise<void> {
   await new Promise(jest.requireActual('timers').setImmediate);
 }
+
+/**
+ * It's common when writing tests to need an object which fits the shape of a
+ * type. However, some properties are unimportant to a test, and so it's useful
+ * if such properties can get filled in with defaults if not explicitly
+ * provided, so that a complete version of that object can still be produced.
+ *
+ * A naive approach to doing this is to define those defaults and then mix them
+ * in with overrides using the spread operator; however, this causes issues if
+ * creating a default value causes a change in global test state — such as
+ * causing a mocked function to get called inadvertently.
+ *
+ * This function solves this problem by allowing defaults to be defined lazily.
+ *
+ * @param defaults - An object where each value is wrapped in a function so that
+ * it doesn't get evaluated unless `overrides` does not contain the key.
+ * @param overrides - The values to override the defaults with.
+ * @param finalizeObject - An optional function to call which will create the
+ * final version of the object. This is useful if you need to customize how a
+ * value receives its default version (say, if it needs be calculated based on
+ * some other property).
+ * @returns The complete version of the object.
+ */
+export function buildTestObject<Type extends Record<PropertyKey, unknown>>(
+  defaults: { [K in keyof Type]: () => Type[K] },
+  overrides: Partial<Type>,
+  finalizeObject?: (object: Type) => Type,
+): Type {
+  const keys = [
+    ...new Set([
+      ...getKnownPropertyNames(defaults),
+      ...getKnownPropertyNames<keyof Type>(overrides),
+    ]),
+  ];
+  const object = keys.reduce((workingObject: Partial<Type>, key) => {
+    if (key in overrides) {
+      return { ...workingObject, [key]: overrides[key] };
+    } else if (key in defaults) {
+      return { ...workingObject, [key]: defaults[key]() };
+    }
+    return workingObject;
+  }, {}) as unknown as Type;
```

Comment:
We can type this `reduce` with a generic parameter instead of using `as unknown as` .

```suggestion
  const object = keys.reduce<Type>((workingObject, key) => {
    if (key in overrides) {
      return { ...workingObject, [key]: overrides[key] };
    } else if (key in defaults) {
      return { ...workingObject, [key]: defaults[key]() };
    }
    return workingObject;
  }, {} as never);
```

Reply:
Hmm, I like typing the "working object" as `Partial<Type>` and then typecasting the final result as `Type` instead of passing `Type` as a type parameter because this makes sense conceptually to me (i.e. it's not complete until it's complete). Now that I say that, I guess I don't need the `as unknown`:

``` suggestion
  const object = keys.reduce((workingObject: Partial<Type>, key) => {
    if (key in overrides) {
      return { ...workingObject, [key]: overrides[key] };
    } else if (key in defaults) {
      return { ...workingObject, [key]: defaults[key]() };
    }
    return workingObject;
  }, {}) as Type;
```

In any case, does this go against any of the guidelines you make in `contributor-docs`?

Reply:
I guess this depends on what we're trying to achieve by applying the `Partial` type, and whether there's a useful difference introduced by using `Partial<Type>` vs. `Type`.

1. `Partial` does not remove any keys from the index signature of `Type`. It only makes the corresponding values nullable (i.e. `| undefined`). 

`Partial<Type>` does not prevent unassigned keys from existing on `workingObject`. Applying `Partial` does not prevent any `key` value from being used for property access on `workingObject` at any point during the `reduce` loop. 

2. `Partial` only enforces that all keys may return `undefined`. It doesn't enforce that some specific set of keys must always return `undefined`. 

If we want to ensure that unassigned keys always return `undefined`, we'd need to write a more accurate but complex type, while keeping track of the iterated and remaining keys each loop e.g.

```ts
type WorkingObject = Omit<Type, IteratedKeys> & Partial<Pick<Type, RemainingKeys>>
```

3. `Partial` expands the probability space rather than helping us narrow it down.

Typing `workingObject` with `Type` restricts its index signature to the correct set of keys, and the assignable property values to the correct set of types. 

Replacing that with `Partial<Type>` expands the range of possibilities to include scenarios where different sets of keys are optional/required, and values are widened to be nullable in addition to whatever type they were previously restricted to.

* * *

I think that the constraints enforced by the `reduce` type parameter is good enough for most purposes, especially if it lets us avoid type assertion. The `@typescript-eslint/prefer-reduce-type-parameter` eslint rule operates on that principle, and it's on our list of rules to re-enable. I believe it could replace the `contributor-docs` entry we have about this topic.

That said, I'd definitely be interested in coming up with a better typing for `reduce` that distinguishes between in-progress and complete states for the working object.

Reply:
Gotcha, thanks for the explanation. It seems that there are two ways to view this:

- You could argue that the working object is an incomplete version of the final version, so `Partial` communicates that it's possible that any of the keys in that object haven't been set yet as it's being built.
- You could also argue that widening the type inside of the `reduce` function is rarely valuable and that it is more valuable to assume that the object is complete all the time.

Despite what I just said above, I guess I don't have a real preference for this. I didn't realize that there was an ESLint rule for preferring the type parameter, so if that's true (and we can use it to remove some words from the contributor docs) then it makes me lean toward your suggestion.